tempesta_fw/tls.c

```diff
@@ -569,6 +570,43 @@ tfw_tls_conn_drop(TfwConn *c)
 	tfw_conn_hook_call(TFW_FSM_HTTP, c, conn_drop);
 }
 
+int tfw_tls_skb_copy_pages(struct sk_buff *skb)
+{
+	int i;
+	struct sk_buff *frag_iter;
+
+	for (i = 0; i < skb_shinfo(skb)->nr_frags; i++) {
+		skb_frag_t *f = &skb_shinfo(skb)->frags[i];
+		struct page *p, *p2;
+		int order;
+		u8 *vaddr, *vaddr2;
+		unsigned int frag_size = skb_frag_size(f);
+
+		p = skb_frag_page(f);
+		order = get_order(frag_size);
+		p2 = alloc_pages(__GFP_MEMALLOC, order);
+		if (!p2)
+			goto err;
+		vaddr = kmap_atomic(p);
+		vaddr2 = kmap_atomic(p2);
+		memcpy(vaddr2, vaddr + f->page_offset, frag_size);
+		kunmap_atomic(vaddr2);
+		kunmap_atomic(vaddr);
```

Comment:
We're in softirq, so there is no preemtion or page faults, so it's safe just to use `page_address()`.